### PR TITLE
feat: add event dispatcher

### PR DESCRIPTION
**Pull request by tanns2**
### Background
The output of AutoGPT is logged to the console and prompt. There is no possibility to directly interact with the output. The aim of the event dispatcher is to provide the possibility for sending the events to a defined endpoint. The endpoint can be any microservice or 3rd party application (which then can do further transformations, e.g. enterprise bus). The event dispatcher fires and forgets all data that is put to the console. It can be seen as the enabler for further integration in a more complex application landscape. 
Note: As this is a data forwarding, a plugin does not make sense as it does not have any effect on the running instance of AutoGPT.

### Changes
The change include a configuration and the event dispatcher itself. The configuration contains all information for the endpoint (protocol, port, host, endpoint) while the event dispatcher just sends the event to the defined endpoint over HTTP POST. The feature is attached to the logs.py typewriter_log method. It forwards the data if the event dispatcher is enabled. The event dispatcher catches every exception that can occur when trying to send the request. It sends the data asynchronosly over a new thread. In case of error and if it is not muted, the error is printed to the console. It never blocks the execution and flow of AutoGPT.

### Documentation
The readme.md contains the configuration documentation. Inside the environments, there is a sample config as well. Further, the code is clean and also documented.

### Test Plan
Tested with following scenarios:

- nothing set up -> nothing is received at endpoint (simple separated HTTP server script)
- set up, not enabled -> nothing is received at endpoint (simple separated HTTP server script)
- set up, enabled, muted -> HTTP script receives events
- set up, no HTTP server running, muted-> AutoGPT works normally, no error output
- set up, no HTTP server running, not muted -> error output in the console
- set up, enabled, wrong configuration, muted -> AutoGPT works normally, no error output
- set up, enabled, wrong configuration, not muted -> error output in the console

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -527,6 +527,21 @@ HUGGINGFACE_API_TOKEN="YOUR_HUGGINGFACE_API_TOKEN"
...
+
+```bash
+EVENT_DISPATCHER_ENABLED=True
+EVENT_DISPATCHER_PROTOCOL=http
+EVENT_DISPATCHER_HOST=localhost
+EVENT_DISPATCHER_ENDPOINT=/other-microservice
+EVENT_DISPATCHER_PORT=45000
+EVENT_DISPATCHER_MUTE=True
+```
+
```
Comment (ntindle) [inline README.md]: the readme.md got refactored significantly. May want to migrate this to a more appropriate location once updated

File: autogpt/config/config.py [hunk 2]
```diff
@@ -136,6 +136,16 @@ def __init__(self) -> None:
...
+        # event dispatcher configuration
+        self.event_dispatcher_enabled = os.getenv("EVENT_DISPATCHER_ENABLED", "False")
+        self.event_dispatcher_protocol = os.getenv("EVENT_DISPATCHER_PROTOCOL")
+        self.event_dispatcher_host = os.getenv(
+            "EVENT_DISPATCHER_HOST",
+        )
+        self.event_dispatcher_endpoint = os.getenv("EVENT_DISPATCHER_ENDPOINT")
+        self.event_dispatcher_port = os.getenv("EVENT_DISPATCHER_PORT")
+        self.event_dispatcher_mute = os.getenv("EVENT_DISPATCHER_MUTE", "True")
+
```
Comment (ntindle) [inline autogpt/config/config.py]: Are defaults needed for the rest of these

File: autogpt/event_dispatcher.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        try:
+            # send the request
+            req = urllib.request.Request(
+                url, data=json_payload, headers=headers, method="POST"
+            )
+            with urllib.request.urlopen(req) as response:
+                pass  # do nothing with the response
+        except Exception as e:  # ensure catching all exceptions
+            if CFG.event_dispatcher_mute == "False":
+                print(f"Error: {e}")
```
Comment (ntindle) [inline autogpt/event_dispatcher.py]: Use the existing debug flag, don't have your own

File: autogpt/event_dispatcher.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    - None
+    """
+
+    json_payload = json.dumps(data).encode("utf-8")
+
+    # Callback definition for thread
+    def send_request():
+        try:
+            # send the request
+            req = urllib.request.Request(
```
Comment (ntindle) [inline autogpt/event_dispatcher.py:39]: Log this on debug flag set

**Review by ntindle — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (tanns2) [reply README.md]: moved

File: autogpt/config/config.py (hunk 2 above)
Comment (tanns2) [reply autogpt/config/config.py]: set

File: autogpt/event_dispatcher.py (hunk 3 above)
Comment (tanns2) [reply autogpt/event_dispatcher.py]: changed to debug flag

File: autogpt/event_dispatcher.py (hunk 4 above)
Comment (tanns2) [reply autogpt/event_dispatcher.py:39]: logged

**Comment by tanns2** [comment]
Thanks for the input, I've added your requested changes.

**Review by ntindle — APPROVED** [review]

**Comment by ntindle** [comment]
fix merge conflicts and its gtg

**Comment by tanns2** [comment]
Hi @ntindle thanks a lot! Fixed, rebased, retested. Looking forward for the merging.

File: autogpt/event_dispatcher.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+                url, data=json_payload, headers=headers, method="POST"
+            )
+            if CFG.debug_mode:
+                print(f"[Event Dispatcher] Info: Sending {req}")
+            with urllib.request.urlopen(req) as response:
+                pass  # do nothing with the response
+        except Exception as e:  # ensure catching all exceptions
+            if CFG.debug_mode:
+                print(f"[Event Dispatcher] Error: {e}")
```
Comment (ntindle) [inline autogpt/event_dispatcher.py]: Use shared logging library for logging

**Review by ntindle — CHANGES_REQUESTED** [review]

**Comment by tanns2** [comment]
@ntindle good point, thanks. Retested and ready for review.

**Comment by tanns2** [comment]
Hi @ntindle or other maintainers, 
I've rebased and pushed again everything to ensure the branch is up to date with a clean commit.
Looking forward for this PR to be merged.

**Comment by p-i-** [comment]
This is a mass message from the AutoGPT core team.
Our apologies for the ongoing delay in processing PRs.
This is because we are re-architecting the AutoGPT core!

For more details (and for infor on joining our Discord), please refer to:
  https://github.com/Significant-Gravitas/Auto-GPT/wiki/Architecting

**Comment by Pwuts** [comment]
@ntindle bump, still wanna merge this? Will need a bit of conflict resolution but nothing too bad.

**Comment by Pwuts** [comment]
Two remarks:
1. The custom `Logger` class has been replaced by the native `logging` module. The same thing can now be achieved by implementing a custom `Handler`, just like we did with the `TTSHandler`.
2. The proposed functionality should be covered by the Agent Protocol afaik, for which support should be added soon. :)

**Comment by Pwuts** [comment]
Closing this in favor of Agent Protocol implementation:
* #5178

**Pull request closed without merging**